conductor-client/src/main/kotlin/com/openlattice/postgres/external/ExternalDatabasePermissioner.kt

```diff
@@ -0,0 +1,515 @@
+import com.openlattice.assembler.PostgresRoles
+import com.openlattice.authorization.*
+import com.openlattice.authorization.processors.GetOrCreateExternalRoleNameEntryProcessor
+import com.openlattice.authorization.securable.SecurableObjectType
+import com.openlattice.edm.EdmConstants
+import com.openlattice.edm.PropertyTypeIdFqn
+import com.openlattice.edm.processors.GetOrganizationIdFromEntitySetEntryProcessor
+import com.openlattice.hazelcast.HazelcastMap
+import com.openlattice.organization.OrganizationExternalDatabaseColumn
+import com.openlattice.organization.OrganizationExternalDatabaseTable
+import com.openlattice.organization.roles.Role
+import com.openlattice.postgres.DataTables
+import com.openlattice.postgres.PostgresPrivileges
+import com.openlattice.postgres.TableColumn
+import com.openlattice.postgres.mapstores.SecurableObjectTypeMapstore
+import com.openlattice.transporter.grantUsageOnSchemaSql
+import com.zaxxer.hikari.HikariDataSource
+import org.apache.olingo.commons.api.edm.FullQualifiedName
+import org.slf4j.LoggerFactory
+import org.springframework.stereotype.Component
+import java.sql.Statement
+import java.util.*
+
+/**
+ * @author Drew Bailey (drew@openlattice.com)
+ */
+@Component
+class ExternalDatabasePermissioner(
+        hazelcastInstance: HazelcastInstance,
+        private val extDbManager: ExternalDatabaseConnectionManager,
+        private val dbCredentialService: DbCredentialService,
+        private val principalsMapManager: PrincipalsMapManager
+): ExternalDatabasePermissioningService {
+    private val atlas: HikariDataSource = extDbManager.connect("postgres")
+
+    private val entitySets = HazelcastMap.ENTITY_SETS.getMap(hazelcastInstance)
+    private val securableObjectTypes = HazelcastMap.SECURABLE_OBJECT_TYPES.getMap(hazelcastInstance)
+    private val organizationExternalDatabaseColumns = HazelcastMap.ORGANIZATION_EXTERNAL_DATABASE_COLUMN.getMap(hazelcastInstance)
+    private val externalRoleNames = HazelcastMap.EXTERNAL_ROLES.getMap(hazelcastInstance)
+
+    companion object {
+        private val logger = LoggerFactory.getLogger(ExternalDatabasePermissioner::class.java)
+
+        private val olToPostgres = mapOf<Permission, Set<PostgresPrivileges>>(
+                Permission.READ to EnumSet.of(PostgresPrivileges.SELECT),
+                Permission.WRITE to EnumSet.of(PostgresPrivileges.INSERT, PostgresPrivileges.UPDATE),
+                Permission.OWNER to EnumSet.of(PostgresPrivileges.ALL)
+        )
+
+        private val allViewPermissions = setOf(Permission.READ)
+
+        private val allTablePermissions = setOf(Permission.READ, Permission.WRITE, Permission.OWNER)
+
+        private fun fqnToColumnName(fqn: FullQualifiedName): String {
+            return fqn.toString()
+        }
+    }
+
+    internal fun configureRolesInDatabase(dataSource: HikariDataSource) {
+        val roles = principalsMapManager.getAllRoles()
+
+        if (roles.isNotEmpty()) {
```

Comment:
someone wise once suggested to me unnesting by doing a `if (isEmpty) { return }` where possible...

Reply:
🙏